task_processing/plugins/kubernetes/kubernetes_pod_executor.py

```diff
@@ -1,20 +1,72 @@
+import threading
+import time
+from enum import auto
+from enum import unique
+from typing import Tuple
+
+from pyrsistent import field
+from pyrsistent import pmap
+from pyrsistent import PRecord
+from pyrsistent import PVector
+from pyrsistent import pvector
+from pyrsistent.typing import PMap
+from pyrsistent.typing import PVector as PVectorType
+
 from task_processing.interfaces import TaskExecutor
 from task_processing.plugins.kubernetes.task_config import KubernetesTaskConfig
+from task_processing.utils import AutoEnum
+
+
+@unique
+class KubernetesTaskState(AutoEnum):
+    TASK_PENDING = auto()
+
+
+class KubernetesTaskMetadata(PRecord):
+    # what box this task/Pod was scheduled onto
+    node_name: str = field(type=str, initial='')
+
+    # the config used to launch this task/Pod
+    task_config: KubernetesTaskConfig = field(type=KubernetesTaskConfig, mandatory=True)
+
+    # TODO(TASKPROC-241): add current task state and task state history as we did for mesos
+    task_state: KubernetesTaskState = field(type=KubernetesTaskState, mandatory=True)
+    # Map of state to when that state was entered (stored as a timestamp)
+    task_state_history: PVectorType[Tuple[KubernetesTaskState, int]] = field(
+        type=PVector, factory=pvector, mandatory=True)
 
 
 class KubernetesPodExecutor(TaskExecutor):
     TASK_CONFIG_INTERFACE = KubernetesTaskConfig
 
-    def run(self, task_config):
-        pass
+    def __init__(self) -> None:
+        self.task_metadata: PMap[str, KubernetesTaskMetadata] = pmap()
+        self._lock = threading.RLock()
+
+    def run(self, task_config: KubernetesTaskConfig) -> None:
+        # we need to lock here since there will be other threads updating this metadata in response
+        # to k8s events
+        with self._lock:
+            self.task_metadata = self.task_metadata.set(
+                task_config.pod_name,
+                KubernetesTaskMetadata(
+                    task_config=task_config,
+                    task_state=KubernetesTaskState.TASK_PENDING,
+                    task_state_history=pvector(
```

Comment:
you actually want to use `v` function here, as `pvector` will turn any sequence type (such as tuple) into PVector rather than forming a PVector with that sequence as first element:

    >>> from pyrsistent import pvector, v
    >>> t = (1,2,)
    >>> pvector(t)
    pvector([1, 2])
    >>> v(t)
    pvector([(1, 2)])